File: python/2022/day7/main.py
tree = {}
curr_dir = []
res = 0
sizes = []


def recurse(path):
    global res
    total = sum([int(x) if x.isnumeric() else recurse(f"{path}/{x}") for x in tree[path]])
    res += total < 100000 and total
    return total


def recurse2(path):
    sizes.append(total := sum([int(x) if x.isnumeric() else recurse2(f"{path}/{x}") for x in tree[path]]))
    return total


for line in [x.strip("\n").split(" ") for x in open("python/2022/day7/test.txt").readlines() if not x.startswith("$ ls")]:
    if "".join(line).startswith("$cd"):
        curr_dir.pop() if line[2] == ".." else curr_dir.append(line[2])
    else:
        if "/".join(curr_dir) in tree:
            tree["/".join(curr_dir)].append(line[1] if line[0] == "dir" else line[0])
        else:
            tree["/".join(curr_dir)] = [line[1] if line[0] == "dir" else line[0]]
        

recurse("/")
print(res)
free_space = recurse2("/") - 40000000
print(min(list(filter(lambda x: (x > free_space), sizes))))

File: typescript/2022/day7/main.ts
import * as fs from "fs";

const tree = {};
const curr_dir: string[] = [];
const sizes: number[] = [];
let res: number = 0;

const recurse = (path: string) => {
  const lst = tree[path];
  let total = 0;
  lst.forEach((element) => {
    if (isNaN(element)) {
      total += recurse(`${path}/${element}`);
    } else {
      total += Number(element);
    }
  });
  res += total < 100000 ? total : 0;
  return total;
};

const recurse2 = (path: string) => {
  const lst = tree[path];
  let total = 0;
  lst.forEach((element) => {
    if (isNaN(element)) {
      total += recurse2(`${path}/${element}`);
    } else {
      total += Number(element);
    }
  });
  sizes.push(total);
  return total;
};

const data = fs
  .readFileSync("typescript/2022/day7/test.txt", "utf-8")
  .split("\r\n")
  .filter((line) => !line.startsWith("$ ls"))
  .forEach((element) => {
    const pair: string[] = element.split(" ");
    if (element.startsWith("$ cd")) {
      pair[2] === ".." ? curr_dir.pop() : curr_dir.push(pair[2]);
    } else {
      if (curr_dir.join("/") in tree) {
        tree[curr_dir.join("/")].push(pair[0] === "dir" ? pair[1] : pair[0]);
      } else {
        tree[curr_dir.join("/")] = [pair[0] == "dir" ? pair[1] : pair[0]];
      }
    }
  });
recurse("/");
console.log(res);

const free_space = recurse2("/") - 40000000;
console.log(Math.min(...sizes.filter((x) => x > free_space)));
